Docs on hooks/actions/filters/events (#9)
alleyinteractive/mantle-docs: features/hooks.md

# Hooks and Events

[[toc]]

WordPress' hooks are the most flexible way of integrating with WordPress. Mantle
aims to make these a bit more flexible for the 21st century. Mantle provides a
simple observer pattern implementation, allowing you to listen for various

> events that occur withing WordPress and Mantle.

Reviewer: ```suggestion
events that occur within WordPress and Mantle.
```

	 * @return void
	 */
	public function handle_order_received( Order_Received $event ) {
		//
	}

	/**
	 * Handle the event at 20 priority.
	 *
	 * @param Order_Cancellation $event
	 * @return void
	 */
	public function handle_handle_cancellation_at_20( Order_Cancellation $event ) {
		//
	}
}
```

Listeners can also automatically listen for WordPress events:

```php
use WP_Query;

class Customer_Listener {
	/**
	 * Handle the query.
	 *
	 * @param WP_Query $event
	 * @return void
	 */
	public function on_pre_get_posts( WP_Query $query ) {
		//
	}

	/**
	 * Handle the redirect.
	 *
	 * @param mixed redirect
	 * @return void
	 */
	public function on_parse_redirect_at_20( $redirect ) {
		//
	}
}
```

#### Caching Discovery

Event discovery is enabled by default and can be disabled via the
`should_discover_events()` method in `Event_Service_Provider`. For performance,
it is ideal to cache the events that are discovered since the Reflection service
is used. To cache the events, run the following command on deployment:

	wp mantle event:cache
## Using Hooks Safely with Type-hint Declarations

WordPress' actions/filters are extended by Mantle to allow for safe use of

> type/return declarations with a fatal error. Mantle providers a wrapper on-top

Reviewer: ```suggestion
type/return declarations with a fatal error. Mantle providers a wrapper on top
```

of `add_action()` and `add_filter()` (these can be used interchangeable with
`Event::listen( ... )`).

### Problem With Type Declarations in Core

Here's an example of a hook with a type-hint that will throw a fatal error (and
potentially bring down your whole site!):

```php
use WP_Post;

add_filter(
  'my_custom_filter',
  function ( array $posts ): array {
    return array_map( fn ( WP_Post $post ) => $post->ID, $posts );
  },
);

> // Else where in your application a plugin adds this filter at a slightly higher priority:

Reviewer: ```suggestion
// Elsewhere in your application a plugin adds this filter at a slightly higher priority:
```

By default, Mantle will attempt to translate type declarations without any need
for customization to prevent errors.

For example, an action is listened for that expects a `string` and has a  `int`
passed. Mantle will properly translate it to a `string`. Type declaring a
`Collection` will have an array automatically translate to a collection and so
on. For the edge cases where a method type declares an object and needs manual
intervention, Mantle will fire a filter that will allow the translation to be
properly handled:

```php
add_filter( 'mantle-typehint-resolve:Custom_Class', function ( $param ) {
	return new Custom_Class( $param );
} );
```

## Find a Hook's Usage in the Codebase

Quickly calculate the usage of a specific WordPress hook throughout your code
base. Mantle will read all specified files in a specific path to find all uses
of a specific action/filter along with their respective line number.

On initial scan of the file system, the results can be a bit slow to build a
cache of all files on the site. By default Mantle will ignore all `test` and
`vendor/` files. The default search path is the `wp-content/` folder of your
installation.

> 	wp mantle hook-usage <hook> [--search-path] [--format]

Reviewer: this is awesome!
